### PR TITLE
Issue #57 invoice dao

### LINKED ISSUE #57: Create class InvoiceDao

### PR DESCRIPTION
**Pull request by nikaalin**

### CODE REVIEW

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+import java.util.List;
+
+public class InvoiceDao implements BasicDaoInterface<InvoiceEntity> {
+    static final String GET_ALL_QUERY = "SELECT * FROM INVOICE";
+    static final String GET_QUERY = "SELECT * FROM INVOICE WHERE ID = ?";
+    static final String SAVE_QUERY = "INSERT INTO INVOICE (ORDER_ID, ISSUE_DATE, " +
+            "PAYMENT_DATE, TOTAL_PRICE, STATUS) VALUES ( ?, ?, ?, ?, ?::e_status_invoice)";
+    static final String UPDATE_QUERY = "UPDATE INVOICE SET (ORDER_ID, ISSUE_DATE, " +
+            "PAYMENT_DATE, TOTAL_PRICE, STATUS) = (?, ?, ?, ?, ?::e_status_invoice) WHERE ID = ?";
+    static final String DELETE_QUERY = "DELETE FROM INVOICE WHERE ID = ?";
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: these could be private?

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        Date issueDate = resultSet.getDate("issue_date");
+        Date paymentDate = resultSet.getDate("payment_date");
+        BigDecimal totalPrice = resultSet.getBigDecimal("total_price");
+        String statusString = resultSet.getString("status").toUpperCase();
+        InvoiceStatus status = InvoiceStatus.valueOf(statusString);
+
+        return new InvoiceEntity(id, orderId, issueDate, paymentDate, totalPrice, status);
+    }
+
+    private void setInvoiceFieldsToStatement(InvoiceEntity invoice, PreparedStatement statement) throws SQLException {
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Too long line

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+        }
+    }
+
+    @Override
+    public InvoiceEntity get(int id) {
+        InvoiceEntity invoice;
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        Connection connection = dbConnector.getConnection();
+        try {
+            PreparedStatement statement = connection.prepareStatement(GET_QUERY, ResultSet.TYPE_SCROLL_SENSITIVE,
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Long line

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Override
+    public void update(InvoiceEntity invoice) {
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        Connection connection = dbConnector.getConnection();
+        try {
+            PreparedStatement statement = connection.prepareStatement(UPDATE_QUERY, ResultSet.TYPE_SCROLL_SENSITIVE,
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Long line

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void delete(InvoiceEntity invoice) {
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        Connection connection = dbConnector.getConnection();
+        try {
+            PreparedStatement statement = connection.prepareStatement(DELETE_QUERY, ResultSet.TYPE_SCROLL_SENSITIVE,
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Long line

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 6]
```diff
@@ -0,0 +1,130 @@
...
+            e.printStackTrace();
+        }
+    }
+
+    @Override
+    public void save(InvoiceEntity invoice) {
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        Connection connection = dbConnector.getConnection();
+        try {
+            PreparedStatement statement = connection.prepareStatement(SAVE_QUERY, ResultSet.TYPE_SCROLL_SENSITIVE,
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Long line

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 7]
```diff
@@ -0,0 +1,130 @@
...
+                    ResultSet.CONCUR_READ_ONLY);
+            statement.setInt(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            resultSet.first();
+            invoice = parseRow(resultSet);
+            connection.close();
+            return invoice;
+        } catch (SQLException e) {
+            e.printStackTrace();
+            throw new RuntimeException(e);
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Throw some kind of checked exception instead (e.g. it's very likely we'll want to know if there's no entity with such id and handle it gracefully)

**Review by a-deriabin — CHANGES_REQUESTED** [review]

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 8]
```diff
@@ -0,0 +1,130 @@
...
+        try {
+            Statement statement = connection.createStatement(ResultSet.TYPE_SCROLL_SENSITIVE,
+                    ResultSet.CONCUR_READ_ONLY);
+            ResultSet resultSet = statement.executeQuery(GET_ALL_QUERY);
+            resultSet.first();
+            do {
+                InvoiceEntity currentInvoice = parseRow(resultSet);
+                invoices.add(currentInvoice);
+            }
+            while (resultSet.next());
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: Check documentation on ResultSet.next().
TL;DR by default, result set points at nothing. Correct usage is `while (resultSet.next()) { ... }`, NOT do/while
This way:
1. You don't need to call resultSet.first() before
2. Code won't fail if you have 0 entries in the result set

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 9]
```diff
@@ -0,0 +1,130 @@
...
+        DBConnectorInterface dbConnector = DBConnectorPostgres.getInstance();
+        Connection connection = dbConnector.getConnection();
+        try {
+            PreparedStatement statement = connection.prepareStatement(GET_QUERY, ResultSet.TYPE_SCROLL_SENSITIVE,
+                    ResultSet.CONCUR_READ_ONLY);
+            statement.setInt(1, id);
+            ResultSet resultSet = statement.executeQuery();
+            resultSet.first();
+            invoice = parseRow(resultSet);
+            connection.close();
```
Comment (a-deriabin) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java]: What happens if SQLException is thrown before connection is closed?

**Review by a-deriabin — APPROVED** [review]

**Review by JustRevolt — APPROVED** [review]

File: src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java [hunk 10]
```diff
@@ -0,0 +1,27 @@
+package com.ifmo.epampractice.dao;
+
+import com.ifmo.epampractice.entity.InvoiceEntity;
+import com.ifmo.epampractice.enums.InvoiceStatus;
+
+import java.math.BigDecimal;
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/dao/InvoiceDao.java:5]: Do you realy need all this imports?

File: src/main/java/com/ifmo/epampractice/daoImpl/InvoiceDaoImpl.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+public class InvoiceDaoImpl implements InvoiceDao {
+    private static final String GET_ALL_QUERY = "SELECT * FROM INVOICE";
+    private static final String GET_QUERY = "SELECT * FROM INVOICE WHERE ID = ?";
+    private static final String SAVE_QUERY = "INSERT INTO INVOICE (ORDER_ID, ISSUE_DATE, " +
+            "PAYMENT_DATE, TOTAL_PRICE, STATUS) VALUES ( ?, ?, ?, ?, ?::e_status_invoice)";
+    private static final String UPDATE_QUERY = "UPDATE INVOICE SET (ORDER_ID, ISSUE_DATE, " +
+            "PAYMENT_DATE, TOTAL_PRICE, STATUS) = (?, ?, ?, ?, ?::e_status_invoice) WHERE ID = ?";
+    private static final String DELETE_QUERY = "DELETE FROM INVOICE WHERE ID = ?";
+
+    private static final Logger log = LogManager.getLogger(InvoiceDaoImpl.class);
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoImpl/InvoiceDaoImpl.java]: It is a constant

File: src/main/java/com/ifmo/epampractice/daoImpl/InvoiceDaoImpl.java [hunk 12]
```diff
@@ -0,0 +1,124 @@
+package com.ifmo.epampractice.daoImpl;
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoImpl/InvoiceDaoImpl.java]: package com.ifmo.epampractice.daoimpl;

File: src/main/java/com/ifmo/epampractice/daoImpl/InvoiceDaoImpl.java [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+                     ResultSet.CONCUR_READ_ONLY);
+             ResultSet resultSet = statement.executeQuery(GET_ALL_QUERY)) {
+            while (resultSet.next()) {
+                InvoiceEntity currentInvoice = parseRow(resultSet);
+                invoices.add(currentInvoice);
+            }
+            return invoices;
+        } catch (SQLException e) {
+            log.error(e);
+            throw new RuntimeException(e);
```
Comment (PavelEzhkov) [inline src/main/java/com/ifmo/epampractice/daoImpl/InvoiceDaoImpl.java]: I think logger will be enough for us, because it is not a good idea to throw a generic exception

**Review by PavelEzhkov — COMMENTED** [review]
fix it

**Review by PavelEzhkov — COMMENTED** [review]
the same as in pull request #69
